### PR TITLE
at least be able to decode stream with multiple IDRs

### PR DESCRIPTION
**Pull request by rwillenbacher**
Your current handling of IDRs and multiple SPS is somewhat off I think.

I tried to change your code so that on the app layer it looks ahead in the bitstream and does a flush when a new IDR is encountered. I had to change the picture buffer flush in picman for IDR pictures as it did not do everything needed to free reference picture. Another approach would be to do a complete teardown and reinit on IDR but that would be quite some speed hit, it would fix certain other issues though.

The decoder then crashes on the 17th SPS  encountered in the stream because there is no ctx->sps_id bounds checking. Might I suggest you reconsider your architecture regarding context and sequence, storing sequence contexts with for example PPS, picture buffer etc, below some sequence structure which is bound to some SPS ID ? Or something like that ?

In case you need a sample stream with multiple ( identical )SPS and IDRs and some closed GOPs of different sizes I uploaded a baseline profile sample here which I think should be correct:
[Some Google Drive Folder](https://drive.google.com/drive/folders/1plFE_-kqDrWs_rqYaHMkVWqxgQyCvQvo?usp=sharing)

Thanks.

### CODE REVIEW

File: app/xevd_app.c [hunk 1]
```diff
@@ -473,16 +476,62 @@ int main(int argc, const char **argv)
...
+                    bs_size = read_bitstream( fp_bs, &i_la_read_pos, bs_buf );
+                    if( bs_size <= 0 )
+                    {
+                        break;
+                    }
+                    i_la_read_pos += XEVD_NAL_UNIT_LENGTH_BYTE + bs_size;
+
+                    nal_unit_type = ( ( bs_buf[ 0 ] >> 1 ) & 0x3f ) - 1;
+
+                    if( nal_unit_type == 1 )
```
Comment (mamthan) [inline app/xevd_app.c:504]: if( nal_unit_type == XEVD_NUT_IDR ) instead of 1, for readability.

File: src_main/xevdm.c [hunk 2]
```diff
@@ -2779,7 +2779,14 @@ int xevd_dec_nalu(XEVD_CTX * ctx, XEVD_BITB * bitb, XEVD_STAT * stat)
...
     {
-        ret = xevdm_eco_sps(bs, sps);
+        XEVD_SPS s_sps;
+
+        memset( &s_sps, 0, sizeof( s_sps ) ) ;
+        ret = xevdm_eco_sps(bs, &s_sps);
+
+        ctx->sps_array[ s_sps.sps_seq_parameter_set_id % 16 ] = s_sps;
+        sps = &ctx->sps_array[ s_sps.sps_seq_parameter_set_id % 16 ];
+        ctx->sps_id = s_sps.sps_seq_parameter_set_id;
```
Comment (mamthan) [inline src_main/xevdm.c:2789]: sps_seq_parameter_set_id ranges from 0 to 15. Module % 16 not required.

File: src_main/xevdm.c [hunk 3]
```diff
@@ -2793,13 +2800,14 @@ int xevd_dec_nalu(XEVD_CTX * ctx, XEVD_BITB * bitb, XEVD_STAT * stat)
...
-        ctx->sps_id++;
+        //ctx->sps_id++; no
     }
     else if (nalu->nal_unit_type_plus1 - 1 == XEVD_NUT_PPS)
     {
         ret = xevdm_eco_pps(bs, sps, pps);
         xevd_assert_rv(XEVD_SUCCEEDED(ret), ret);
-        ctx->sps = &ctx->sps_array[pps->pps_seq_parameter_set_id];
+        ctx->sps_id = pps->pps_seq_parameter_set_id % 16;
+        sps = ctx->sps = &ctx->sps_array[ ctx->sps_id ];
```
Comment (mamthan) [inline src_main/xevdm.c:2810]: sps_id should not be based on pps_seq_parameter_set_id. Please check.

Comment (mamthan) [reply src_main/xevdm.c]: yes, ctx->sps_count is unused in the current xevd code. 
So, then there is no need to check ctx->sps_id is out of bounds. Also ctx->sps_id is based on sps->sps_seq_parameter_set_id in current xevd code.

**Review by mamthan — APPROVED** [review]
Sps bound check modifications to be reverted back, based on the discussion. Looks good to me.

**Comment by rwillenbacher** [comment]
So looking at "Files changed" all that this PR is doing now is the Flush on IDR in the App and the more careful picture buffer flush (picman_flush_pb) for baseline and main profile. Merge ? :>

**Comment by mss-park** [comment]
> So looking at "Files changed" all that this PR is doing now is the Flush on IDR in the App and the more careful picture buffer flush (picman_flush_pb) for baseline and main profile. Merge ? :>

Thanks for your update to solve the problem of Multiple IDR.
But the changing Application can be affect to other project. Thus, other method was tested to solve the problem in #26 .
Instead of flushing all remaining pictures in DPB, the POC number of that pictures is changed. And the number of POC to be outputted is changed to the number of minimum POC in DPB.
The bitstreams you shared were tested, and it seems to be working.
I referred to your code to implement the modified code. Thank you again.
And please check the updated code there is other problem in Multiple IDR case.

**Comment by kpchoi** [comment]
@rwillenbacher 

Hi, I think https://github.com/mpeg5/xevd/pull/26 solved this issue.
Could you check the code? it is merged to master branch.

**Comment by rwillenbacher** [comment]
@kpchoi 

I gave the master branch a try and the number of frames output now matches the number of pictures in the stream. The decoder also does not get stuck in some endless loop on certain picture buffer operations anymore. However, in the YUV output some IDR pictures are switched with the picture before them, this cannot be detected with the picture HASH as these are picture local. I can be detected by comparing the output if my pull request with the output of the current master though.

Anyway I think the wrong picture order is because in master there is code like this in xevd_picman.c::picman_flush_pb();
```    /* mark all frames unused */
    int max_poc = 0;
    for(i = 0; i < MAXM_PB_SIZE; i++)
    {
        if (pm->pic[i] && IS_REF(pm->pic[i]))
        {
            SET_REF_UNMARK(pm->pic[i]);
            if (pm->pic[i] && pm->pic[i]->poc > max_poc)
            {
                max_poc = pm->pic[i]->poc;
            }
            picman_move_pic(pm, i, MAX_PB_SIZE - 1);
            i--;
        }
    }
    max_poc = max_poc == 0 ? max_poc : max_poc + 1;
```

So only the max poc of pictures which are still reference frames when this is called are taken into consideration. Now if the last picture of the gop before the IDR is no reference picture it most likely has the highest POC but max poc is set to something smaller, so the poc gets adjusted to >= 0 in the loop below that code, leading to duplicate POCs around 0 or for POC 0.

I think it has to be something like this:

```
    /* mark all frames unused */
    int max_poc = 0;
    for (i = 0; i < MAX_PB_SIZE; i++)
    {
        if( pm->pic[ i ] && pm->pic[ i ]->need_for_out && pm->pic[ i ]->poc > max_poc )
        {
            max_poc = pm->pic[ i ]->poc;
        }
        if( pm->pic[ i ] && IS_REF( pm->pic[ i ] ) )
        {
            SET_REF_UNMARK(pm->pic[i]);
            picman_move_pic(pm, i, MAX_PB_SIZE - 1);
            i--;
        }
    }
    max_poc = max_poc == 0 ? max_poc : max_poc + 1;
```

Then IDR pictures are not switched around, for me anyway.

Sadly I do not have the EVC standard in writing yet. So I am unable to look up the relevant mechanics and just taking guesses.

**Comment by mss-park** [comment]
@rwillenbacher 
Thank you for updating the problem. It was fixed in #27.
And  Encoder side multiple IDR was updated in (https://github.com/mpeg5/xeve/pull/49)
Multiple-IDR can be tested using updated encoder and decoder.
Please let us know, if you find any problem.

**Comment by rwillenbacher** [comment]
Well, as XEVD now works with streams from my encoder everything is ok I guess, so I'll close this PR.

As a Note, I am closing a GOP by truncating/not encoding non present pictures which somewhat un-intuitive but it seems to work for Baseline where POCs are derived instead of coded. I wonder if there is a better way which allows having the last picture of a GOP as reference, like shorten the GOP instead of truncating... ah well, does not matter.

**Comment by mpeg5** [comment]
@rwillenbacher 
Thanks for your efforts.
We could find the problem and fix it because of your comments.
Hopefully, create PR if you find some better way regarding IDR decoding.
Thanks again.

**Pull request closed without merging**